### PR TITLE
Add JumpToNext/PreviousBreakpoint

### PR DESCRIPTION
**Pull request by dkaszews**
Tested automatically on vim and NeoVim using the written test methods.

Half of #633, will add `JumpToPC` in separate PR to keep them small (less implementation, more tests as need to let PC move about).

### CODE REVIEW

**Comment by puremourning** [comment]
This change is [Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)

File: autoload/vimspector.vim [hunk 1]
```diff
@@ -469,6 +469,20 @@ function! vimspector#JumpToBreakpointViewBreakpoint() abort
   py3 _vimspector_session.JumpToBreakpointViewBreakpoint()
 endfunction
 
+function! vimspector#JumpToNextBreakpoint() abort
```
Comment (dkaszews) [inline autoload/vimspector.vim:472]: New issue: This file is almost entirely composed of such forwarding functions. They could be generated from a list instead.

**Comment by puremourning** [comment]
Nice thanks. 

One this worth testing is what happens when the server moves a breakpoint - ie put a bp on a var decl, start debugging and check where cursor goes. There is a distinction in the code between the user's requested position and the actual position the server used.

**Review by puremourning — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @dkaszews and @puremourning)

___
*[`autoload/vimspector.vim` line 472 at r1](https://reviewable.io/reviews/puremourning/vimspector/639#-N8PSxGVQqsXVDKqaVN0-r1-472:-N8PvrPOABCtoRaU-CZU:b9jtrwn) ([raw file](https://github.com/puremourning/vimspector/blob/e205ae0c0506ddb1d5195d47899d29b071113fec/autoload/vimspector.vim#L472)):*
<details><summary><i>Previously, dkaszews wrote…</i></summary><blockquote>

New issue: This file is almost entirely composed of such forwarding functions. They could be generated from a list instead.
</blockquote></details>

Actually I prefer it this way. It’s explicit and makes it grepable.

**Comment by dkaszews** [comment]
> > _[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)_ status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @dkaszews and @puremourning)
> 
> _[`autoload/vimspector.vim` line 472 at r1](https://reviewable.io/reviews/puremourning/vimspector/639#-N8PSxGVQqsXVDKqaVN0-r1-472:-N8PvrPOABCtoRaU-CZU:b9jtrwn) ([raw file](https://github.com/puremourning/vimspector/blob/e205ae0c0506ddb1d5195d47899d29b071113fec/autoload/vimspector.vim#L472)):_
> 
> Previously, dkaszews wrote…
> > New issue: This file is almost entirely composed of such forwarding functions. They could be generated from a list instead.
> 
> Actually I prefer it this way. It’s explicit and makes it grepable.

Understand the logic of not wanting fully generated functions, but what about slightly shortening it with a helper? Something like: `InvokePyIfEnabled` to make it a oneliner.

**Comment by puremourning** [comment]
> > > _[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)_ status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @dkaszews and @puremourning)
> 
> > 
> 
> > _[`autoload/vimspector.vim` line 472 at r1](https://reviewable.io/reviews/puremourning/vimspector/639#-N8PSxGVQqsXVDKqaVN0-r1-472:-N8PvrPOABCtoRaU-CZU:b9jtrwn) ([raw file](https://github.com/puremourning/vimspector/blob/e205ae0c0506ddb1d5195d47899d29b071113fec/autoload/vimspector.vim#L472)):_
> 
> > 
> 
> > Previously, dkaszews wrote…
> 
> > > New issue: This file is almost entirely composed of such forwarding functions. They could be generated from a list instead.
> 
> > 
> 
> > Actually I prefer it this way. It’s explicit and makes it grepable.
> 
> 
> 
> Understand the logic of not wanting fully generated functions, but what about slightly shortening it with a helper? Something like: `InvokePyIfEnabled` to make it a oneliner. 

I just don't see the point. It's not like we add enough of these to warrant metaprogramming.

**Comment by dkaszews** [comment]
Accidentally pushed changes to wrong branch, rolled back with `git push -f`

**Comment by puremourning** [comment]
PythonLint CI is failing (you can run locally with `make check`)

**Review by puremourning — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @dkaszews and @puremourning)

*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eDdd4Dle5zBOZd4Nz:-N8eDdd50zJJ6NQWQBJH:b-2gzq81) (no related file):*
We should add `<Plug>` mappings for these functions 

___
*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eDh-g3LWKElAUPgGf:-N8eDh-g3LWKElAUPgGg:b7re2pm) (no related file):*
We need to update `README.md`

___
*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eELP1B3QB6wC2exZP:-N8eELP1B3QB6wC2exZQ:b5xiq47) (no related file):*
need to add a test which does the jumping after a breakpoint is moved by the server to ensure we're using the right line number (i.e. the adjusted line number)

___
___
*[`python3/vimspector/breakpoints.py` line 270 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eE-M974ePwgnfFee8:-N8eE-MAEVgRmMjP2O-w:b-22ovj5) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/python3/vimspector/breakpoints.py#L270)):*
> ```Python
>     self.UpdateUI()
> 
>   def JumpToBreakpoint( self, bp ):
> ```

this should probably be a "internal" method now `_JumpToBreakpoint`

indeed it probably doesn't need to be a member anymore as it doesn't appear to use `self` anywhere, so we can promote this to a module scope function to make that distinction clear



_Code quote:_
```Python
JumpToBreakpoint
```
___
*[`python3/vimspector/breakpoints.py` line 306 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eEdrRDV9SUxTSlIUW:-N8eEdrRDV9SUxTSlIUX:b-unulvw) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/python3/vimspector/breakpoints.py#L306)):*
> ```Python
>       ( bp
>         for bp in sorted( bps, key=operator.itemgetter( 'lnum' ), reverse=reverse )
>         if comparator( bp[ 'lnum' ], line ) ),
> ```

I don't think that we can just use `bp[ 'line' ]` here as it will be "wrong" if the server moves/adjusts the line when verifying.

you can see elsewhere we have this sort of stuff:

```
        line = bp[ 'line' ]
        if 'server_bp' in bp:
          server_bp = bp[ 'server_bp' ]
          line = server_bp.get( 'line', line )
```

probably time for a `_GetBreakpointLine( bp )` rather than this

**Review by puremourning — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @dkaszews and @puremourning)

___
*[`tests/breakpoints.test.vim` line 953 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eF8SD5-PxxqlSYF8s:-N8eF8SD5-PxxqlSYF8t:bs6jch1) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/tests/breakpoints.test.vim#L953)):*
> ```vim
>   lcd testdata/cpp/simple
>   edit simple.cpp
>   let main_win_id = win_getid()
> ```

this doesn't appear to be used



_Code quote:_
```vim
main_win_id
```

**Review by puremourning — COMMENTED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: all files reviewed, 7 unresolved discussions (waiting on @dkaszews and @puremourning)

**Comment by dkaszews** [comment]
> I don't think that we can just use bp[ 'line' ] here as it will be "wrong" if the server moves/adjusts the line when verifying.

Will write a test for it, but I think it should work as it uses breakpoints window which I believe automatically updates.

**Review by dkaszews — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: all files reviewed, 6 unresolved discussions (waiting on @puremourning)

*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eELP1B3QB6wC2exZP:-N92jhD1-00CfpeCsdbT:bww4rec) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

need to add a test which does the jumping after a breakpoint is moved by the server to ensure we're using the right line number (i.e. the adjusted line number)
</blockquote></details>

Added test, verified it works.

___
*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eDdd4Dle5zBOZd4Nz:-N92lvvp0V7_pIJWzpq9:bb6un1j) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

We should add `<Plug>` mappings for these functions 
</blockquote></details>

Done

___
*[a discussion](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eDh-g3LWKElAUPgGf:-N92lxFW9Ea9PfAvmfec:b-i2kwd) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

We need to update `README.md`
</blockquote></details>

Added to mappings table

___
___
*[`python3/vimspector/breakpoints.py` line 270 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eE-M974ePwgnfFee8:-N92m1ep0yAcv7xLhA48:bqs0h67) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/python3/vimspector/breakpoints.py#L270)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

this should probably be a "internal" method now `_JumpToBreakpoint`

indeed it probably doesn't need to be a member anymore as it doesn't appear to use `self` anywhere, so we can promote this to a module scope function to make that distinction clear
</blockquote></details>

Moved to module scope, prefixed with `_`
___
*[`python3/vimspector/breakpoints.py` line 306 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eEdrRDV9SUxTSlIUW:-N92jmInEZGghOnQTmsD:b-6lminm) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/python3/vimspector/breakpoints.py#L306)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I don't think that we can just use `bp[ 'line' ]` here as it will be "wrong" if the server moves/adjusts the line when verifying.

you can see elsewhere we have this sort of stuff:

```
        line = bp[ 'line' ]
        if 'server_bp' in bp:
          server_bp = bp[ 'server_bp' ]
          line = server_bp.get( 'line', line )
```

probably time for a `_GetBreakpointLine( bp )` rather than this 
</blockquote></details>

Test shows it works, because the breakpoint is taken from automatically updated breakpoints window immediatelly before jump
___
*[`tests/breakpoints.test.vim` line 953 at r2](https://reviewable.io/reviews/puremourning/vimspector/639#-N8eF8SD5-PxxqlSYF8s:-N92k0u0DwzsJHSd71sq:ba3dfob) ([raw file](https://github.com/puremourning/vimspector/blob/ebb38c130e93cc4db3b8146993f54b3dd8ffecd2/tests/breakpoints.test.vim#L953)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

this doesn't appear to be used
</blockquote></details>

Removed

**Review by puremourning — APPROVED** [review]
:lgtm:

Thanks!

> Reviewed 22 of 22 files at r4.
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639#-:-N92pVwI889TVYSAQoGW:b9benhj)* status: 6 of 23 files reviewed, 1 unresolved discussion (waiting on @puremourning)

**Review by puremourning — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 6 of 23 files reviewed, 1 unresolved discussion (waiting on @dkaszews and @puremourning)

___
*[`tests/breakpoints.test.vim` line 1010 at r6](https://reviewable.io/reviews/puremourning/vimspector/639#-N92xHqo09FE0T4bcWxy:-N92xHqo09FE0T4bcWxz:b89y2d8) ([raw file](https://github.com/puremourning/vimspector/blob/3dffed554926a3cd46e91d5aea96778908b65d3e/tests/breakpoints.test.vim#L1010)):*
> ```vim
>   " First assert is needed to wait for launch to finish before moving cursor
>   call vimspector#Launch()
>   call vimspector#test#signs#AssertCursorIsAtLineInBuffer( 'simple.cpp', s:break_main_line, 1 )
> ```

80 characters max. Break at the `,`

ditto elsewhere



_Code quote:_
```vim
  call vimspector#test#signs#AssertCursorIsAtLineInBuffer( 'simple.cpp', s:break_main_line, 1 )
```

**Review by puremourning — COMMENTED** [review]
> Reviewed 16 of 17 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: all files reviewed, 1 unresolved discussion (waiting on @dkaszews)

**Review by dkaszews — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 22 of 23 files reviewed, 1 unresolved discussion (waiting on @puremourning)

___
*[`tests/breakpoints.test.vim` line 1010 at r6](https://reviewable.io/reviews/puremourning/vimspector/639#-N92xHqo09FE0T4bcWxy:-N92ysrwBS6TbYWWqBiM:b-yjmi9w) ([raw file](https://github.com/puremourning/vimspector/blob/3dffed554926a3cd46e91d5aea96778908b65d3e/tests/breakpoints.test.vim#L1010)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

80 characters max. Break at the `,`

ditto elsewhere
</blockquote></details>

Done, shouldn't it be checked by `VimscriptLint`?

**Review by puremourning — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639)* status: 22 of 23 files reviewed, all discussions resolved (waiting on @puremourning)

___
*[`tests/breakpoints.test.vim` line 1010 at r6](https://reviewable.io/reviews/puremourning/vimspector/639#-N92xHqo09FE0T4bcWxy:-N92zFXH6cSwMJ3Ftt58:b-aa7f9r) ([raw file](https://github.com/puremourning/vimspector/blob/3dffed554926a3cd46e91d5aea96778908b65d3e/tests/breakpoints.test.vim#L1010)):*
<details><summary><i>Previously, dkaszews (Dominik Kaszewski) wrote…</i></summary><blockquote>

Done, shouldn't it be checked by `VimscriptLint`?
</blockquote></details>

`vint` doesn't have this check. I tried to add it myself, but vint's AST seemed to be lacking the info to do this check

**Review by puremourning — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/puremourning/vimspector/639#-:-N92zPqv5k5_fNwQkMLY:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @dkaszews)

**Pull request merged**